src/epd2in13_v3/mod.rs

```diff
@@ -0,0 +1,577 @@
+            self.set_ram_address_counters(spi, x, y)?;
+
+            self.cmd_with_data(spi, Command::WriteRamRed, buffer)?;
+        }
+
+        Ok(())
+    }
+
+    /// Never use directly this function when using partial refresh, or also
+    /// keep the base buffer in syncd using `set_partial_base_buffer` function.
+    fn display_frame(&mut self, spi: &mut SPI, _delay: &mut DELAY) -> Result<(), SPI::Error> {
+        if self.refresh == RefreshLut::Full {
+            self.set_display_update_control_2(
+                spi,
+                DisplayUpdateControl2::new()
+                    .enable_clock()
+                    .enable_analog()
+                    .display()
+                    .disable_analog()
+                    .disable_clock(),
+            )?;
+        } else {
+            self.set_display_update_control_2(spi, DisplayUpdateControl2::new().display())?;
+        }
+        self.command(spi, Command::MasterActivation)?;
+        self.wait_until_idle();
+
+        Ok(())
+    }
+
+    fn update_and_display_frame(
+        &mut self,
+        spi: &mut SPI,
+        buffer: &[u8],
+        delay: &mut DELAY,
+    ) -> Result<(), SPI::Error> {
+        self.update_frame(spi, buffer, delay)?;
+        self.display_frame(spi, delay)?;
+
+        if self.refresh == RefreshLut::Quick {
+            self.set_partial_base_buffer(spi, buffer)?;
+        }
+        Ok(())
+    }
+
+    fn clear_frame(&mut self, spi: &mut SPI, _delay: &mut DELAY) -> Result<(), SPI::Error> {
+        let color = self.background_color.get_byte_value();
+
+        self.set_ram_area(spi, 0, 0, WIDTH - 1, HEIGHT - 1)?;
+        self.set_ram_address_counters(spi, 0, 0)?;
+
+        self.command(spi, Command::WriteRam)?;
+        self.interface.data_x_times(
+            spi,
+            color,
+            buffer_len(WIDTH as usize, HEIGHT as usize) as u32,
+        )?;
+
+        // Always keep the base buffer equals to current if not doing partial refresh.
+        if self.refresh == RefreshLut::Full {
+            self.set_ram_area(spi, 0, 0, WIDTH - 1, HEIGHT - 1)?;
+            self.set_ram_address_counters(spi, 0, 0)?;
+
+            self.command(spi, Command::WriteRamRed)?;
+            self.interface.data_x_times(
+                spi,
+                color,
+                buffer_len(WIDTH as usize, HEIGHT as usize) as u32,
+            )?;
+        }
+        Ok(())
+    }
+
+    fn set_background_color(&mut self, background_color: Color) {
+        self.background_color = background_color;
+    }
+
+    fn background_color(&self) -> &Color {
+        &self.background_color
+    }
+
+    fn width(&self) -> u32 {
+        WIDTH
+    }
+
+    fn height(&self) -> u32 {
+        HEIGHT
+    }
+
+    fn set_lut(
+        &mut self,
+        spi: &mut SPI,
+        refresh_rate: Option<RefreshLut>,
+    ) -> Result<(), SPI::Error> {
+        let buffer = match refresh_rate {
+            Some(RefreshLut::Full) | None => &LUT_FULL_UPDATE,
```

Comment:
Optional: We could merge v2 and v3 if we use a feature gate or sth similar, which would reduce the duplicated code by a lot.

Reply:
I would suggest merging everything into the same structure and just having a different `new` constructor.

If you insist on having 2 different structures, you can add a generic parameter or a const generic to make them different types.

Reply:
Different constructor sounds like a good idea